### PR TITLE
Added max_expansion param to span_multi

### PR DESCRIPTION
**Pull request by nirmalc**
Causes span query to throw exception if term expansions with in
multi_term inner query exceeds max_expansion param

Example query:

{"span_multi" : {"match" : {"prefix" : {"user" : {"value" : "ki","boost" : 1.08}}},"max_expansions" : 22 ,"boost" :1.0}}

I was trying to patch one older version of ES to avoid heap crash, but noticed that its logged as an issue here - https://github.com/elastic/elasticsearch/issues/27432

### CODE REVIEW

**Comment by mayya-sharipova** [comment]
@nirmalc Thank you for submitting the PR.  It nicely addresses  #27432

So this PR, although submitted against master, will be for 6.4 version.
For 7.x version, we will need another PR, slightly different.

So, for 6.x version, we want to do following: 
- [x] Add `max_expansions` into `span_multi` query as an optional parameter that defaults to `1024` 
- [ ]  Issue a deprecation warning if a user submits `span_multi` query without `max_expansions` parameter
- [ ]  throw an error if `max_expansions` is greater than 1024
- [x] add max_expansions as an override of the rewrite in 6.x
- [ ] update documentation for `span_multi` query

Unchecked items still need to be implemented in your PR

File: server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java [hunk 1]
```diff
@@ -117,7 +147,44 @@ public static SpanMultiTermQueryBuilder fromXContent(XContentParser parser) thro
...
+                        "exceeds maxExpansions [maxExpansions is set to " + maxExpansions + "]");
+                }
+            }
+
+            @Override
+            protected void addClause(List<SpanQuery> topLevel, Term term, int docCount, float boost, TermContext states) {
+                SpanTermQuery q = new SpanTermQuery(term, states);
+                topLevel.add(q);
+            }
+        };
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java]: I would not throw an error, I would just ensure that we collect only up to `maxExpansions` queries. @jimczi what do you think?

Comment (nirmalc) [reply server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java]: Thanks , I will look at checklist . I thought the plan was to throw exception if term expansion count is above max_expansions ( which is defaulted to 1024 ) . The other pattern can be done already by adding ```rewrite``` param to inner multi query

**Review by jimczi — CHANGES_REQUESTED** [review]
Thanks @nirmalc. The change looks good but I wonder if we really need a `max_expansion` option or if we should only honor `BooleanQuery.getMaxClauseCount()` ?  MultiTermQuery that uses a TopTermRewrite strategy have the same limit so it would be consistent and wouldn't require the addition of another option that could be confusing since the rewrite method can be overridden in the multi term query directly.

File: server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java (hunk 1 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java]: I think it's better to throw an exception, at least in master. I don't think we need a `max_expansion` option though, we can just honor `BooleanQuery#getMaxClauseCount()` and fails the query if there are more terms. The span scoring rewrite does not allow to limit the number of expansions, it is all or nothing and I think it's a good property since it will not confuse the users by returning partial responses on prefix queries that match more than 1024 terms. This is the main reason why I opened the issue in the first place, we should never build a disjunction on more than `BooleanQuery.getMaxClauseCount()`.

File: server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java [hunk 2]
```diff
@@ -117,7 +147,44 @@ public static SpanMultiTermQueryBuilder fromXContent(XContentParser parser) thro
                     "[span_multi] must have [" + MATCH_FIELD.getPreferredName() + "] multi term query clause");
         }
 
-        return new SpanMultiTermQueryBuilder(subQuery).queryName(queryName).boost(boost);
+        return new SpanMultiTermQueryBuilder(subQuery).queryName(queryName).boost(boost).maxExpansions(maxExpansionValue);
+    }
+
+    public class TopTermSpanBooleanQueryRewriteWithMaxClause extends SpanMultiTermQueryWrapper.SpanRewriteMethod{
+        MultiTermQuery multiTermQuery;
```
[the remark this answers is not in the document]
Comment (nirmalc) [reply server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java]: 👍 , will do  -  I thought it will be nice to have  more dynamic way to override instead of boolean.max count ( I believe its a static property and requires yml file change ) ; Dynamically changing is very useful in case of cluster with multiple indices/data shapes. But I do agree that it can be confusing .

File: server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java [hunk 3]
```diff
@@ -117,7 +147,44 @@ public static SpanMultiTermQueryBuilder fromXContent(XContentParser parser) thro
                     "[span_multi] must have [" + MATCH_FIELD.getPreferredName() + "] multi term query clause");
         }
 
-        return new SpanMultiTermQueryBuilder(subQuery).queryName(queryName).boost(boost);
+        return new SpanMultiTermQueryBuilder(subQuery).queryName(queryName).boost(boost).maxExpansions(maxExpansionValue);
+    }
+
+    public class TopTermSpanBooleanQueryRewriteWithMaxClause extends SpanMultiTermQueryWrapper.SpanRewriteMethod{
```
[the remark this answers is not in the document]
Comment (nirmalc) [reply server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java]: just used in one of unit tests..

File: server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java]: > Dynamically changing is very useful in case of cluster with multiple indices/data shapes. 

Which `multi_term` query are you using ? Is it to increase the default value of 1024 ? Any multi term query that don't use top terms rewriting is not a good fit for the span query in general. If you need to perform prefix queries within span queries it is preferable to index the prefix using the `edge_ngram` filter. This way you can transform any prefix query into a span term query on a single term.

[the remark this answers is not in the document]
Comment (nirmalc) [reply server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java]: Thanks for feedback @jimczi  , Multi queries could be prefix , wildcard or worst case  -regex. Im using span queries now for exact use case mentioned in ES documentation of span queries  - legal discovery. We considered edge ngrams which takes care of part of that use case. data volume( text)  and unpredictable cardinality  is another issue ( often few thousands for many multi searches) 

My request is to consider it as advanced param , but if elastic thinks it is confusing for users - i can change PR to use boolean query limit. Please let me know ! Thanks again

[the remark this answers is not in the document]
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java]: My original intent was to add a `max_expansion` to force the rewrite of all multi terms to use a top terms rewrite when used inside a span query but in this case the `max_expansion` param can be interpreted as the number of top terms to keep during the rewrite. However I prefer your approach which throws an exception on `multi_term` queries that don't use top terms rewrite and matches more than 1024 terms. In this case we can fail the query with a nice message explaining that a top term rewrite should be used on the `multi_term` query if applicable. `max_expansion` is confusing in this scenario and doesn't bring much so I think we should simply honor the BooleanQuery max clause limit.

[the remark this answers is not in the document]
Comment (nirmalc) [reply server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java]: 👍  i will update the PR in a day or two

[the remark this answers is not in the document]
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java]: thanks @nirmalc !

**Comment by nirmalc** [comment]
@jimczi , please let me know if changes look good , I'll squash and rebase with master

File: docs/reference/query-dsl/span-multi-term-query.asciidoc [hunk 4]
```diff
@@ -36,3 +36,8 @@ GET /_search
 }
 --------------------------------------------------
 // CONSOLE
+
+span multi query will hit too many clauses failure if it exceeds the boolean
+query limit (defaults  to `1024`) unless inner multi query uses top_terms_N
+rewrite method.
```
Comment (jimczi) [inline docs/reference/query-dsl/span-multi-term-query.asciidoc]: Maybe add a link to the documentation: <<query-dsl-multi-term-rewrite,rewrite>>

File: server/src/test/java/org/elasticsearch/search/query/SpanMultiQueryIT.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+import org.elasticsearch.test.ESIntegTestCase;
+
+import static org.elasticsearch.cluster.metadata.IndexMetaData.SETTING_NUMBER_OF_SHARDS;
+
+import java.util.ArrayList;
+import java.util.concurrent.ExecutionException;
+
+public class SpanMultiQueryIT extends ESIntegTestCase {
+    public void testTermExpansionExceptionOnSpanFailure() throws ExecutionException, InterruptedException {
+        org.elasticsearch.common.settings.Settings.Builder builder = org.elasticsearch.common.settings.Settings.builder();
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/search/query/SpanMultiQueryIT.java]: You can import Settings and remove the package name ?

File: server/src/test/java/org/elasticsearch/search/query/SpanMultiQueryIT.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        builder.put(SETTING_NUMBER_OF_SHARDS, 1).build();
+
+        createIndex("test", builder.build());
+        ArrayList<IndexRequestBuilder> reqs = new ArrayList<>();
+        for (int i = 0; i < BooleanQuery.getMaxClauseCount() + 1; i++) {
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/search/query/SpanMultiQueryIT.java]: That's a big test to check for an exception. I'd something like:
````
final int origValue = BooleanQuery.getMaxClauseCount();
BooleanQuery.setMaxClauseCount(5);
try {
   // test goes here
} finally {
  BooleanQuery.setMaxClauseCount(origValue);
}

File: server/src/test/java/org/elasticsearch/search/query/SpanMultiQueryIT.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+import org.elasticsearch.index.query.SpanNearQueryBuilder;
+import org.elasticsearch.index.query.SpanTermQueryBuilder;
+import org.elasticsearch.test.ESIntegTestCase;
+
+import static org.elasticsearch.cluster.metadata.IndexMetaData.SETTING_NUMBER_OF_SHARDS;
+
+import java.util.ArrayList;
+import java.util.concurrent.ExecutionException;
+
+public class SpanMultiQueryIT extends ESIntegTestCase {
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/search/query/SpanMultiQueryIT.java]: Can you move these tests to org.elasticsearch.search.query.SearchQueryIT, we don't need another IT test to check this.

File: server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java [hunk 8]
```diff
@@ -114,12 +123,55 @@ public static SpanMultiTermQueryBuilder fromXContent(XContentParser parser) thro
...
+            @Override
+            protected Query build(List<SpanQuery> builder) {
+                return new SpanOrQuery((SpanQuery[]) builder.toArray(new SpanQuery[builder.size()]));
+            }
+
+            @Override
+            protected void checkMaxClauseCount(int count) {
+                if (count > maxExpansions) {
+                    throw new ElasticsearchException("[" + multiTermQuery.toString() + " ] " +
+                        "exceeds maxExpansions [maxExpansions is set to " + maxExpansions + "]");
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java]: Maybe mention boolean query max clause and top terms rewrite ?

**Review by jimczi — COMMENTED** [review]
I left more comments but I like it overall. 
Can you also merge with latest master, I'll run the tests afterward.

**Comment by nirmalc** [comment]
@jimczi new commit should address all code review comments .

File: server/src/test/java/org/elasticsearch/search/query/SearchQueryIT.java [hunk 9]
```diff
@@ -1819,4 +1826,52 @@ public void testRangeQueryRangeFields_24744() throws Exception {
         assertHitCount(searchResponse, 1);
     }
 
+    public void testTermExpansionExceptionOnSpanFailure() throws ExecutionException, InterruptedException {
+        org.elasticsearch.common.settings.Settings.Builder builder = org.elasticsearch.common.settings.Settings.builder();
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/search/query/SearchQueryIT.java]: nit: add an import for Settings

File: server/src/test/java/org/elasticsearch/search/query/SearchQueryIT.java [hunk 10]
```diff
@@ -1819,4 +1826,52 @@ public void testRangeQueryRangeFields_24744() throws Exception {
...
+                    Integer.toString(i) + " bar baz"));
+            }
+            indexRandom(true, false, reqs);
+
+            QueryBuilder queryBuilder = new SpanNearQueryBuilder(new SpanMultiTermQueryBuilder(QueryBuilders.wildcardQuery
+                ("body", "f*")), 0).addClause(new SpanTermQueryBuilder("body", "bar"));
+
+            expectThrows(ElasticsearchException.class, () ->
+                client().prepareSearch().setIndices("test").setQuery(queryBuilder).get());
+        }finally{
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/search/query/SearchQueryIT.java]: nit: missing space

File: server/src/test/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilderTests.java [hunk 11]
```diff
@@ -162,4 +161,95 @@ public void testFromJson() throws IOException {
...
+
+        assertTrue(query instanceof SpanMultiTermQueryWrapper);
+        if (query instanceof SpanMultiTermQueryWrapper) {
+            MultiTermQuery.RewriteMethod rewriteMethod = ((SpanMultiTermQueryWrapper)query).getRewriteMethod();
+            assertTrue(rewriteMethod instanceof SpanMultiTermQueryBuilder.TopTermSpanBooleanQueryRewriteWithMaxClause);
+        }
+
+    }
+
+    public void testDefaultMaxRewriteWithDefaultValue() throws Exception {
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilderTests.java]: This is a duplicate of the previous test ?

File: server/src/test/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilderTests.java [hunk 12]
```diff
@@ -162,4 +161,95 @@ public void testFromJson() throws IOException {
...
+            "\"prefix\" : {" +
+            "\"user\" : {" +
+            "\"value\" : \"ki\"," +
+            "\"boost\" : 1.08" +
+            "}"+
+            "}"+
+            "}," +
+            "\"boost\" :1.0" +
+            "}" +
+            "}";
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilderTests.java]: The xcontent parsing is tested in the parent class, can you modify this test to use the query builders instead ?

**Review by jimczi — COMMENTED** [review]
It looks good @nirmalc ! I left some minor comments but I think it's getting close.

**Comment by jimczi** [comment]
@nirmalc can you merge with master and resolve the conflicts ? I'll trigger the tests when it's done.

**Comment by nirmalc** [comment]
Done, sorry - my fork was bit old and  i had to get rid of some tests since we decided not to do ```max_expansion``` params.

File: docs/reference/query-dsl/span-multi-term-query.asciidoc [hunk 13]
```diff
@@ -37,10 +37,7 @@ GET /_search
...
-terms is large. To avoid an unbounded expansion you can set the
-<<query-dsl-multi-term-rewrite,rewrite method>> of the multi term query to `top_terms_*`
-rewrite. Or, if you use `span_multi` on `prefix` query only, you can
-activate the <<index-prefix-config,`index_prefixes`>> field option of the `text` field instead. This will
-rewrite any prefix query on the field to a a single term query that matches the indexed prefix.
\ No newline at end of file
+span multi query will hit too many clauses failure if it exceeds the boolean
+query limit (defaults  to `1024`) unless inner multi query uses top_terms_N
+<<query-dsl-multi-term-rewrite,rewrite>> parameter.
+
```
Comment (jimczi) [inline docs/reference/query-dsl/span-multi-term-query.asciidoc]: Please keep the WARNING here and maybe reword this to:
````
WARNING: `span_multi` queries will hit too many clauses failure if the number of terms that match the query exceeds the boolean query limit (defaults to 1024). To avoid an unbounded expansion you can set the
<<query-dsl-multi-term-rewrite,rewrite method>> of the multi term query to `top_terms_*`
rewrite. Or, if you use `span_multi` on `prefix` query only, you can
activate the <<index-prefix-config,`index_prefixes`>> field option of the `text` field instead. This will
rewrite any prefix query on the field to a a single term query that matches the indexed prefix.
````

File: server/src/test/java/org/elasticsearch/search/query/SearchQueryIT.java [hunk 14]
```diff
@@ -1819,4 +1826,52 @@ public void testRangeQueryRangeFields_24744() throws Exception {
...
+            expectThrows(ElasticsearchException.class, () ->
+                client().prepareSearch().setIndices("test").setQuery(queryBuilder).get());
+        } finally {
+            BooleanQuery.setMaxClauseCount(origBoolMaxClauseCount);
+        }
+
+    }
+
+    public void testTermExpansionExceptionOnSpan() throws ExecutionException, InterruptedException {
+        Settings.Builder builder = Settings.builder();
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/search/query/SearchQueryIT.java]: I don't think this test is needed. `testSpanMultiTermQuery` does the same thing.

**Review by jimczi — CHANGES_REQUESTED** [review]
Thanks @nirmalc , it's really close however I left some comments that need to be addressed before we run the tests.

**Comment by nirmalc** [comment]
updated doc + removed test you suggested as already there is one testing that scenario.

**Review by jimczi — APPROVED** [review]
LGTM thanks @nirmalc , I'll merge if the tests pass

**Comment by jimczi** [comment]
test this please
